File: practice/algorithms/search/cut_the_tree/cut_the_tree.py
#!/usr/bin/env python

from Queue import Queue

def bfs(i_graph, i_start, i_n):
    visited = set()
    visit_order = []
    parent = [-1 for _ in range(i_n)]
    q = Queue()
    q.put(i_start)
    while not q.empty():
        NODE = q.get()
        if NODE not in visited:
            visited.add(NODE)
            visit_order.append(NODE)
            for VERT in i_graph[NODE]:
                if VERT not in visited:
                    q.put(VERT)
                    parent[VERT] = NODE
    return parent, visit_order

def make_graph(i_edges, i_n):
    r_list = [set([]) for _ in range(i_n)]
    for EDGE in i_edges:
        r_list[EDGE[0]].add(EDGE[1])
        r_list[EDGE[1]].add(EDGE[0])
    return r_list
            
n = int(raw_input())
w = map(int, raw_input().split())

total = sum(w)
edges = []
for _ in range(n - 1):
    v1, v2 = map(int, raw_input().split())
    edges.append((v1 - 1, v2 - 1))

graph = make_graph(edges, n)
parent, visit_order = bfs(graph, 0, n)

sums = [0 for _ in range(n)]
for VERT in reversed(visit_order):
    sums[VERT] += w[VERT]
    c = VERT
    p = parent[c]
    if p != -1:
        sums[p] += sums[VERT]
        
m = total
for CNT in range(n):
    m = min(m, abs(2 * sums[CNT] - total))
    
print m

File: practice/algorithms/search/cut_the_tree/cut_the_tree.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <queue>
#include <set>
#include <numeric>

using namespace std;

pair<vector<int>, vector<int>> bfs(const vector<set<int>> & i_graph, int i_start, int i_size)
{
    vector<bool> visited(i_size, false);
    vector<int> visit_order;
    
    vector<int> parent(i_size, -1);
    
    queue<int> q;
    q.push(i_start);
    
    while (!q.empty())
    {
        int node = q.front();
        q.pop();
        
        if (!visited[node])
        {
            visited[node] = true;
            visit_order.push_back(node);
            
            set<int>::iterator it = i_graph[node].begin();
            for (; it != i_graph[node].end(); ++it)
            {
                if (!visited[*it])
                {
                    q.push(*it);
                    parent[*it] = node;
                }
            }
        }
    }
    
    return make_pair(parent, visit_order);
}

vector<set<int>> make_graph(const vector<pair<int, int>> & i_edges, int i_size)
{
    vector<set<int>> graph(i_size);
    
    vector<pair<int, int>>::const_iterator it = i_edges.begin();
    for (; it != i_edges.end(); ++it)
    {
        graph[it->first].insert(it->second);
        graph[it->second].insert(it->first);
    }
    
    return graph;
}

int main()
{
    int n, v1, v2;
    
    cin >> n;
    
    vector<int> weights(n);
    for (std::size_t pos = 0; pos < n; ++pos)
    {
        cin >> weights[pos];
    }
    
    vector<pair<int, int>> edges;
    for (int cnt = 0; cnt < n - 1; ++cnt)
    {
        cin >> v1 >> v2;
        edges.push_back(make_pair(v1 - 1, v2 - 1));
    }
    
    const int total = accumulate(weights.begin(), weights.end(), 0);
    
    vector<set<int>> graph = make_graph(edges, n);
    
    pair<vector<int>, vector<int>> bfs_res = bfs(graph, 0, n);
    
    vector<int> parent = bfs_res.first;
    vector<int> visit_order = bfs_res.second;
    
    vector<int> sums(n, 0);
    
    vector<int>::reverse_iterator rit = visit_order.rbegin();
    for (; rit != visit_order.rend(); ++rit)
    {
        sums[*rit] += weights[*rit];
        int p = parent[*rit];
        if (p != -1)
        {
            sums[p] += sums[*rit];
        }
    }
    
    int res = total;
    for (int cnt = 0; cnt < n; ++cnt)
    {
        res = min(res, abs(2 * sums[cnt] - total));
    }
    
    cout << res;
    
    return 0;
}

